File: android/app/src/main/kotlin/com/stockbook/app/feature/today/EarningsPdf.kt
package com.stockbook.app.feature.today

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Typeface
import android.graphics.pdf.PdfDocument
import com.stockbook.core.text.EarningsDocument
import java.io.File

/**
 * Draws an [EarningsDocument] onto one A4 page.
 *
 * [SummaryPdf]'s geometry — A4 at 72dpi, the same margin, the same rules — so
 * every page the shop prints looks like it came out of the same shop. One page
 * always: the chain is six lines and the confession is two, and a summary that
 * ran over would not be one.
 *
 * **Black on white**, like the statement, and for a different reason. That one is
 * handed to a customer; this one is never handed to anyone — but it is printed,
 * and a dark page prints badly whoever is reading it.
 *
 * The layout decides nothing about wording: every string it draws came from
 * [EarningsDocument], which is shared with the iOS build and tested.
 */
object EarningsPdf {

    private const val PAGE_WIDTH = 595
    private const val PAGE_HEIGHT = 842
    private const val MARGIN = 44f

    private const val TITLE_SIZE = 15f
    private const val BODY_SIZE = 9.5f
    private const val ROW_SIZE = 10f
    private const val LINE = 13f
    private const val BAND_HEIGHT = 19f

    private fun paint(size: Float, bold: Boolean = false, grey: Boolean = false) = Paint().apply {
        isAntiAlias = true
        textSize = size
        color = if (grey) 0xFF6B6B76.toInt() else 0xFF14141C.toInt()
        typeface = Typeface.create(Typeface.SANS_SERIF, if (bold) Typeface.BOLD else Typeface.NORMAL)
    }

    private val rule = Paint().apply {
        isAntiAlias = true
        color = 0xFFD6D6DE.toInt()
        strokeWidth = 0.8f
    }

    private val band = Paint().apply {
        isAntiAlias = true
        color = 0xFFEDEDF2.toInt()
    }

    fun write(document: EarningsDocument, into: Context, fileName: String): File {
        val pdf = PdfDocument()
        val page = pdf.startPage(PdfDocument.PageInfo.Builder(PAGE_WIDTH, PAGE_HEIGHT, 1).create())
        val canvas = page.canvas
        var y = MARGIN

        val right = PAGE_WIDTH - MARGIN

        // --- Whose figures these are, and which days

        canvas.drawText(document.shopName, MARGIN, y + BODY_SIZE, paint(BODY_SIZE, grey = true))
        y += LINE + 4
        canvas.drawText(document.title, MARGIN, y + TITLE_SIZE, paint(TITLE_SIZE, bold = true))
        y += TITLE_SIZE + 6
        canvas.drawText(document.onDate, MARGIN, y + BODY_SIZE, paint(BODY_SIZE, grey = true))
        y += LINE + 20

        if (document.isEmpty) {
            canvas.drawText(document.emptyLine, MARGIN, y + BODY_SIZE, paint(BODY_SIZE))
            pdf.finishPage(page)
            return pdf.saveTo(into, fileName)
        }

        // --- Takings down to what was kept

        for (line in document.lines) {
            val isTotal = line.weight == EarningsDocument.Weight.TOTAL
            val isMinus = line.weight == EarningsDocument.Weight.MINUS

            // A total is what the lines above it come to, so the rule goes over
            // it — the same place a subtotal's rule goes on every other page
            // this shop prints.
            if (isTotal) {
                y += 4
                canvas.drawLine(MARGIN, y, right, y, rule)
                y += 9
            }

            val body = paint(ROW_SIZE, bold = isTotal, grey = isMinus && !isTotal)
            canvas.drawText(line.label, MARGIN, y + ROW_SIZE, body)
            canvas.drawTextRight(if (isMinus) "− ${line.value}" else line.value, right, y + ROW_SIZE, body)
            y += if (isTotal) 20 else 16
        }

        // --- What the page could not account for

        if (document.hasGap) {
            y += 14
            canvas.drawRect(MARGIN, y, right, y + BAND_HEIGHT, band)
            canvas.drawText(
                document.gapHeading.uppercase(),
                MARGIN + 7f,
                y + BAND_HEIGHT - 6.5f,
                paint(BODY_SIZE, bold = true)
            )
            y += BAND_HEIGHT + 10

            for (line in document.gap) {
                canvas.drawText(line.label, MARGIN, y + BODY_SIZE, paint(BODY_SIZE, grey = true))
                canvas.drawTextRight(line.value, right, y + BODY_SIZE, paint(BODY_SIZE, grey = true))
                y += 14
            }

            document.gapNote?.let {
                y += 4
                canvas.drawText(it, MARGIN, y + BODY_SIZE, paint(BODY_SIZE, grey = true))
            }
        }

        pdf.finishPage(page)
        return pdf.saveTo(into, fileName)
    }

    private fun PdfDocument.saveTo(context: Context, fileName: String): File {
        val file = File(context.cacheDir, fileName)
        file.outputStream().use { writeTo(it) }
        close()
        return file
    }

    private fun Canvas.drawTextRight(text: String, rightEdge: Float, y: Float, paint: Paint) {
        drawText(text, rightEdge - paint.measureText(text), y, paint)
    }
}

File: ios/Stockbook/Features/Today/EarningsPDF.swift
import UIKit

/// Draws an `EarningsDocument` onto one A4 page.
///
/// `SummaryPDF`'s geometry — A4 at 72dpi, the same margin, the same rules — so
/// every page the shop prints looks like it came out of the same shop. One page
/// always: the chain is six lines and the confession is two, and a summary that
/// ran over would not be one.
///
/// **Black on white**, like the statement, and for a different reason. That one
/// is handed to a customer; this one is never handed to anyone — but it is
/// printed, and a dark page prints badly whoever is reading it.
///
/// The layout decides nothing about wording. Every string it draws came from
/// `EarningsDocument`, which is shared with the Android build and tested.
enum EarningsPDF {

    private static let pageSize = CGSize(width: 595, height: 842)
    private static let margin: CGFloat = 44

    private static let titleSize: CGFloat = 15
    private static let bodySize: CGFloat = 9.5
    private static let rowSize: CGFloat = 10
    private static let line: CGFloat = 13
    private static let bandHeight: CGFloat = 19

    private static let ink = UIColor(red: 0.078, green: 0.078, blue: 0.110, alpha: 1)
    private static let grey = UIColor(red: 0.420, green: 0.420, blue: 0.463, alpha: 1)
    private static let ruleColour = UIColor(red: 0.839, green: 0.839, blue: 0.871, alpha: 1)
    private static let bandColour = UIColor(red: 0.929, green: 0.929, blue: 0.949, alpha: 1)

    private static func attributes(_ size: CGFloat, bold: Bool = false, muted: Bool = false) -> [NSAttributedString.Key: Any] {
        [
            .font: bold ? UIFont.boldSystemFont(ofSize: size) : UIFont.systemFont(ofSize: size),
            .foregroundColor: muted ? grey : ink
        ]
    }

    static func write(_ document: EarningsDocument, fileName: String) throws -> URL {
        let renderer = UIGraphicsPDFRenderer(bounds: CGRect(origin: .zero, size: pageSize))
        let url = FileManager.default.temporaryDirectory.appendingPathComponent(fileName)

        try renderer.writePDF(to: url) { context in
            context.beginPage()

            let right = pageSize.width - margin
            var y = margin

            // MARK: Whose figures these are, and which days

            document.shopName.draw(at: CGPoint(x: margin, y: y), withAttributes: attributes(bodySize, muted: true))
            y += line + 4
            document.title.draw(at: CGPoint(x: margin, y: y), withAttributes: attributes(titleSize, bold: true))
            y += titleSize + 6
            document.onDate.draw(at: CGPoint(x: margin, y: y), withAttributes: attributes(bodySize, muted: true))
            y += line + 20

            guard !document.isEmpty else {
                document.emptyLine.draw(at: CGPoint(x: margin, y: y), withAttributes: attributes(bodySize))
                return
            }

            // MARK: Takings down to what was kept

            for figure in document.lines {
                let isTotal = figure.weight == .total
                let isMinus = figure.weight == .minus

                // A total is what the lines above it come to, so the rule goes
                // over it — the same place a subtotal's rule goes on every other
                // page this shop prints.
                if isTotal {
                    y += 4
                    rule(from: CGPoint(x: margin, y: y), to: CGPoint(x: right, y: y))
                    y += 9
                }

                let style = attributes(rowSize, bold: isTotal, muted: isMinus && !isTotal)
                figure.label.draw(at: CGPoint(x: margin, y: y), withAttributes: style)
                drawRight(isMinus ? "− \(figure.value)" : figure.value, rightEdge: right, y: y, attributes: style)
                y += isTotal ? 20 : 16
            }

            // MARK: What the page could not account for

            if document.hasGap {
                y += 14
                bandColour.setFill()
                UIBezierPath(rect: CGRect(x: margin, y: y, width: right - margin, height: bandHeight)).fill()
                document.gapHeading.uppercased()
                    .draw(at: CGPoint(x: margin + 7, y: y + 4), withAttributes: attributes(bodySize, bold: true))
                y += bandHeight + 10

                for entry in document.gap {
                    entry.label.draw(at: CGPoint(x: margin, y: y), withAttributes: attributes(bodySize, muted: true))
                    drawRight(entry.value, rightEdge: right, y: y, attributes: attributes(bodySize, muted: true))
                    y += 14
                }

                if let note = document.gapNote {
                    y += 4
                    note.draw(at: CGPoint(x: margin, y: y), withAttributes: attributes(bodySize, muted: true))
                }
            }
        }

        return url
    }

    private static func drawRight(_ text: String, rightEdge: CGFloat, y: CGFloat, attributes: [NSAttributedString.Key: Any]) {
        let width = (text as NSString).size(withAttributes: attributes).width
        text.draw(at: CGPoint(x: rightEdge - width, y: y), withAttributes: attributes)
    }

    private static func rule(from: CGPoint, to: CGPoint) {
        let path = UIBezierPath()
        path.move(to: from)
        path.addLine(to: to)
        path.lineWidth = 0.8
        ruleColour.setStroke()
        path.stroke()
    }
}
